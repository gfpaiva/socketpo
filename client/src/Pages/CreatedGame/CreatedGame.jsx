import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import { Link } from 'react-router-dom';

import Button from '../../Components/Button/Button';

import './CreatedGame.scss';

const CreatedGame = ({ createdGame }) => {

	const [ copied, updateCopied ] = useState(false);

	let copyTimeout = null;
	const gameUrl = `${window.location.protocol}//${window.location.host}/game/${createdGame.hash}`;

	const copyHandler = e => {
		e.preventDefault();

		const copyFunc = e => {
			e.preventDefault();
			e.clipboardData && e.clipboardData.setData("text/plain", gameUrl);
		};

		document.addEventListener('copy', copyFunc, false);
		document.execCommand('copy');
		document.removeEventListener('copy', copyFunc, false);

		updateCopied(true);
		copyTimeout = setTimeout(() => {
			updateCopied(false);
		}, 2000);
	};

	// Clear timeout anyway on unmount
	useEffect(() => {
		return () => {
			clearTimeout(copyTimeout);
		}
	});

	return (
		<div className="page page--created">
			<h2>Game has been created: </h2>
			<p><strong>Name: </strong>{createdGame.name}</p>
			<p>
				<strong>Link: </strong>
				<span className="created__game-url">{gameUrl}</span>
				<Button
					spaced
					onClick={copyHandler}
				>
					{copied ? 'Link Copied' : 'Copy Link'}
				</Button>
			</p>

			<p><Link className='link  link--underline' to={`/game/${createdGame.hash}`}>Go to the game</Link></p>
		</div>
	);
};

CreatedGame.propTypes = {
	createdGame: PropTypes.object.isRequired
};

export default CreatedGame;

