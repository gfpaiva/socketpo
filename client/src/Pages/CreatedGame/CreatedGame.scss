.created__game {
	&-url {
		padding: 0 .5em;
		word-break: break-all;
	}
}
