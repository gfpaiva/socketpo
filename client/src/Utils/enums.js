import React from 'react';
import Avatar from 'avataaars';

import {
	Rock,
	Paper,
	Scissors,
} from '../Components/Icons/Icons';

const statusValues = {
	0: 'Stand By',
	1: 'In progress',
	2: 'Finished'
};

const playValues = {
	0: 'Miss Play',
	1: 'Rock',
	2: 'Paper',
	3: 'Scissors'
};

const playIcons = {
	1: <Rock />,
	2: <Paper />,
	3: <Scissors />
}

const avatarValues = {
	1: <Avatar
		topType='LongHairMiaWallace'
		accessoriesType='Prescription02'
		hairColor='BrownDark'
		facialHairType='Blank'
		clotheType='Hoodie'
		clotheColor='PastelBlue'
		eyeType='Happy'
		eyebrowType='Default'
		mouthType='Smile'
		skinColor='Light'
	/>,
	2: <Avatar
		topType='LongHairBigHair'
		accessoriesType='Blank'
		hairColor='Blonde'
		facialHairType='Blank'
		clotheType='ShirtVNeck'
		clotheColor='PastelYellow'
		eyeType='Wink'
		eyebrowType='RaisedExcitedNatural'
		mouthType='Serious'
		skinColor='DarkBrown'
	/>,
	3: <Avatar
		topType='ShortHairShortFlat'
		accessoriesType='Prescription01'
		hairColor='Blonde'
		facialHairType='Blank'
		clotheType='BlazerSweater'
		eyeType='Default'
		eyebrowType='DefaultNatural'
		mouthType='Serious'
		skinColor='Pale'
	/>,
	4: <Avatar
		topType='ShortHairShortWaved'
		accessoriesType='Blank'
		hairColor='Auburn'
		facialHairType='BeardLight'
		facialHairColor='Red'
		clotheType='BlazerShirt'
		clotheColor='Gray01'
		eyeType='WinkWacky'
		eyebrowType='UnibrowNatural'
		mouthType='Grimace'
		skinColor='DarkBrown'
	/>,
};

export const parseStatus = statusNum => statusValues[statusNum];
export const parsePlay = playNum => playValues[playNum];
export const parsePlayIcons = playNum => playIcons[playNum];
export const parseAvatar = avatarNum => avatarValues[avatarNum];
