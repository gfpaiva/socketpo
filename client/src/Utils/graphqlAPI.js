import gql from 'graphql-tag';

export const totalGameFields = `
name
hash
status
players {
	id
	name
	avatar
	ready
	roundsWin
}
results {
	message
	matchWinner {
		id
		name
		avatar
	}
	rounds {
		isDraw
		finished
		plays {
			player {
				id
				name
			}
			play
		}
		winner {
			id
			name
			avatar
		}
	}
}`;

export const getGame = gql`
	query getGame($hash: String!) {
		GameByHash(hash: $hash) {
			${totalGameFields}
		}
	}
`;

export const play = gql`
	mutation play($hash: String!, $player: PlayerInput!, $play: Int! ) {
		play(hash: $hash, player: $player, play: $play ) {
			hash
			status
		}
	}
`;

export const gameSub = gql`
	subscription gameSub($hash: String!) {
		gameSubscription(hash: $hash) {
			game {
				${totalGameFields}
			}
			message
			hash
			player
		}
	}
`;
