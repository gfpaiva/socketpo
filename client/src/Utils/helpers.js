export const sanitize = str => (str.toLowerCase()
.replace(/[àáâãäå]/ig, 'a')
.replace(/[éèëê]/ig, 'e')
.replace(/[íìïî]/ig, 'i')
.replace(/[óòõöô]/ig, 'o')
.replace(/[uúùüû]/ig, 'u')
.replace(/[ç]/ig, 'c')
.replace(/^[0-9]/ig, '')
.replace(/\s/ig, '-'));
