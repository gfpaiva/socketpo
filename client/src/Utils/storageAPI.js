export const setObject = (key, object) => localStorage.setItem(key, JSON.stringify(object));
export const getObject = key => JSON.parse(localStorage.getItem(key));
