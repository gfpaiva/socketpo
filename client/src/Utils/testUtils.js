export const fakePlayer = (rewrite) => ({
	__typename: 'Player',
	avatar: 3,
	id: 'player01',
	name: 'Player 01',
	ready: false,
	roundsWin: 0,
	...rewrite
});

export const fakeGame = (fullPlayers = true, matchWinner = null, rounds = [], rewrite) => {

	const players = fullPlayers ? [fakePlayer(), fakePlayer({ id: 'player02', name: 'Player 02' })] : [fakePlayer()]

	return {
		__typename: 'Game',
		id: 'game01',
		hash: 'game01',
		name: 'Game 01',
		players,
		results: {
			__typename: "Result",
			matchWinner,
			message: null,
			rounds
		},
		status: 0,
		...rewrite
	};
};

export const fakeRound = (rewrite) => ({
	__typename: 'Round',
	isDraw: false,
	finished: true,
	plays: [
		fakePlay({ play: 1 }),
		fakePlay({ player: fakePlayer({ id: 'player02', name: 'Player 02' }), play: 3 })
	],
	winner: fakePlayer(),
	...rewrite
});

export const fakePlay = (rewrite) => ({
	__typename: 'Play',
	player: fakePlayer(),
	play: 0,
	...rewrite
});
