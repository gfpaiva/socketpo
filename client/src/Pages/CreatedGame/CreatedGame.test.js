import React from 'react';
import { mount } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom';
import CreatedGame from './CreatedGame';
import { fakeGame } from '../../Utils/testUtils';

const game = fakeGame(false);
global.document.execCommand = jest.fn()

describe('<CreatedGame />', () => {

	let wrapper;

	beforeEach(() => {
		wrapper = mount(
			<Router>
				<CreatedGame
					createdGame={game}
				/>
			</Router>
		);
	})

	it('should mount properly', () => {

		expect(wrapper.find('CreatedGame')).toMatchSnapshot();
	});

	it('should handle copy properly', () => {

		wrapper.find('Button').simulate('click');

		expect(wrapper.find('CreatedGame').state().copied).toBeTruthy();
	});
});
