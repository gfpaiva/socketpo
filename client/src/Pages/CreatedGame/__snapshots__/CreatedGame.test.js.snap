// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreatedGame /> should mount properly 1`] = `
<CreatedGame
  createdGame={
    Object {
      "__typename": "Game",
      "hash": "game01",
      "id": "game01",
      "name": "Game 01",
      "players": Array [
        Object {
          "__typename": "Player",
          "avatar": 3,
          "id": "player01",
          "name": "Player 01",
          "ready": false,
          "roundsWin": 0,
        },
      ],
      "results": Object {
        "__typename": "Result",
        "matchWinner": null,
        "message": null,
        "rounds": Array [],
      },
      "status": 0,
    }
  }
>
  <div
    className="page page--created"
  >
    <h2>
      Game has been created: 
    </h2>
    <p>
      <strong>
        Name: 
      </strong>
      Game 01
    </p>
    <p>
      <strong>
        Link: 
      </strong>
      <span
        className="created__game-url"
      >
        http://localhost/game/game01
      </span>
      <Button
        onClick={[Function]}
        spaced={true}
      >
        <button
          className="
			button
			
			
			 button--spaced
			
			"
          onClick={[Function]}
        >
          Copy Link
        </button>
      </Button>
    </p>
    <p>
      <Link
        className="link  link--underline"
        to="/game/game01"
      >
        <a
          className="link  link--underline"
          href="/game/game01"
          onClick={[Function]}
        >
          Go to the game
        </a>
      </Link>
    </p>
  </div>
</CreatedGame>
`;
